### ISSUE TITLE
commitBuffer: invalid argument (invalid character)

### ISSUE THREAD

**Issue by akvadrako** [Gabriel439/bench#40]
Using the latest `nixos/nix` docker image and `nix-env -i bench`, I can't even benchmark `true`:

```
477495791d20:/app# bench true
benchmarking true
time                 453.2 bench: <stdout>: commitBuffer: invalid argument (invalid character)
```

**Comment by Gabriel439** [comment]
Yeah, that is very weird.  Usually I associate that error message with a locale misconfiguration when the program attempts to display a Unicode character outside of the ASCII character set.  However, neither `bench` nor `true` display any such characters.

I will take a stab at reproducing this on my end to see if I can narrow down what is going on.  In the meantime, could you include the output of the following command run from inside the container:

```
$ nix-instantiate --eval --expr 'builtins.readFile <nixpkgs/.git-revision>'
```

... so I can know exactly what revision of `nixpkgs` that your container had installed

**Comment by akvadrako** [comment]
I have tried with a few nix versions, but this is one example:

```sh
cb8929afb58d:/app# nix-instantiate --eval --expr 'builtins.readFile <nixpkgs/.git-revision>'
"d91c70171fb96ee42f10b2d447430b8a9cc576a8"
```

And the Dockerfile:

```docker
FROM nixos/nix:2.0.4
RUN nix-env -i bench
CMD bench true
```

**Comment by Gabriel439** [comment]
This appears to be an issue in the upstream `criterion` library.  See https://github.com/bos/criterion/issues/205 for more details.

There is a work-around on my end that I can apply but I want to see if it can be fixed upstream first

**Comment by Gabriel439** [comment]
@akvadrako: Alright, this is fixed in the upstream `criterion` and `code-page` libraries.  At some point those changes will make their way into Nixpkgs and then the `bench` you install using `nix-env` will correctly display Unicode characters.

In the meantime, you might be able to work around this locally by setting `LANG=en_US.UTF-8` or something similar

**Comment by mwu** [comment]
I'm not sure if this is still a problem for bench but it was for nixos/nix and tasty and I ended up here when googling this error. The correct LANG value is C.UTF-8 and it can't be anything else (at time of writing anyway) when using a nixos/nix image.

The root cause is:
* nixos/nix is based on alpine, which uses musl, which doesn't have complete locale support (see https://wiki.musl-libc.org/functional-differences-from-glibc.html#Character_sets_and_locale)
* nixos/nix only setting LANG in `/etc/profile`, which only takes effect for login shells

**Comment by dcastro** [comment]
Just ran into this issue (using nix and tasty, just like @mwu), and I thought I'd add my solution here.

Wrapping `main` with [`withUtf8`](https://hackage.haskell.org/package/with-utf8-1.0.2.1/docs/Main-Utf8.html#v:withUtf8) seems to work.

```hs
-- before
main :: IO ()
main =
  tests >>= nettestMainWithIngredients ourIngredients

-- after
main :: IO ()
main =
  withUtf8 $
    tests >>= nettestMainWithIngredients ourIngredients
```
